src/rle.rs

```diff
@@ -0,0 +1,398 @@
+            return Ok(());
+        }
+
+        if self.byte == byte {
+            self.reps += 1;
+            return Ok(())
+        }
+
+        if self.byte != byte {
+            try!(self.flush());
+            self.reps = 1;
+            self.byte = byte;
+        }
+
+        Ok(())
+    }
+}
+
+impl<W: Write> Write for Encoder<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut bytes_written = 0;
+
+        for byte in buf {
+            try!(self.process_byte(*byte));
+            bytes_written += 1;
+        }
+
+        Ok(bytes_written)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        if self.reps == 1 {
+            try!(self.w.write(&[self.byte]));
+        } else if self.reps > 1 {
+            let mut buf = [0; 11];
+            let mut reps_encode = self.reps - 2;
+            buf[0] = self.byte;
+            buf[1] = self.byte;
+
+            let mut index = 2;
+
+            loop {
+                buf[index] = (reps_encode & 0b0111_1111) as u8;
+                reps_encode = reps_encode >> 7;
+
+                if reps_encode == 0 {
+                    buf[index] = buf[index] | 0b1000_0000;
+                    break;
+                }
+
+                index += 1;
+            }
+
+            try!(self.w.write(&buf[..(index + 1)]));
+        }
+
+        Ok(())
+    }
+}
+
+struct RunBuilder {
+    byte: u8,
+    slice: [u8; 9],
+    byte_count: u8
+}
+
+impl RunBuilder {
+    fn new(byte: u8) -> RunBuilder {
+        RunBuilder {
+            byte: byte,
+            slice: [0; 9],
+            byte_count: 0
+        }
+    }
+
+    fn to_run(&mut self) -> Run {
+        let reps = 2 + self.slice.iter().enumerate().fold(0u64, |reps, (i, &byte)| {
+            reps | (((byte & 0b0111_1111) as u64) << (i as u32 * 7))
+        });
+
+        Run {
+            byte: self.byte,
+            reps: reps
+        }
+    }
+
+    fn add_byte(&mut self, byte: u8) -> io::Result<()> {
+        if self.byte_count >= 9 {
+            Err(io::Error::new(io::ErrorKind::Other, "Overly long run"))
+        } else {
+            self.slice[self.byte_count as usize] = byte;
+            self.byte_count += 1;
+            Ok(())
+        }
+    }
+}
+ 
+struct Run {
+    byte: u8,
+    reps: u64
+}
+
+enum DecoderState {
+    Clean,
+    Single(u8),
+    Run(RunBuilder)
+}
+
+/// This structure is used to decode a run length encoded stream. This wraps
+/// an internal reader which is read from when this decoder's read method is
+/// called.
+pub struct Decoder<R> {
+    r: R,
+    buf: VecDeque<u8>,
+    state: DecoderState,
+    run: Option<Run>
+}
+
+impl<R: Read> Decoder<R> {
+    /// Creates a new decoder which will read data from the given stream. The
+    /// inner stream can be re-acquired by moving out of the `r` field of this
+    /// structure.
+    pub fn new(r: R) -> Decoder<R> {
+        Decoder {
+            r: r,
+            buf: VecDeque::with_capacity(32),
```

Comment:
Perhaps instead of using a `VecDeque` this could use a `BufReader`? That would maybe reduce the number of overall reads and you could reuse the underlying buffer.